### PR TITLE
Update Waveshare_7_5.ino

### PR DESCRIPTION
**Pull request by hoffmakl1961**
rework of UpdateLocalTime()
>> array for weekday and month, defined in the lang_xx.h and implement a country specific date format.
for DE     "Mo., 24. Juni 2019"

and here the changes in the lang_gr.h. 

[lang_gr.zip](https://github.com/G6EJD/ESP32-e-Paper-Weather-Display/files/3320555/lang_gr.zip)

### CODE REVIEW

File: examples/Waveshare_7_5/Waveshare_7_5.ino [hunk 1]
```diff
@@ -498,58 +499,38 @@ boolean SetupTime() {
...
 boolean UpdateLocalTime() {
   struct tm timeinfo;
-  char   output[30], day_output[30];
-  char   day_number[30], day_year[30];
-  char   day_lang[4], month_lang[4];
-  char   update_time[30];
-  while (!getLocalTime(&timeinfo, 5000)) { // Wait for 5-sec for time to synchronise
+  char   output[23], day_output[30];
+  char   update_time[9];            // change from 30 to 9      
+  while (!getLocalTime(&timeinfo, 5000)) {    // Wait for 5-sec for time to synchronise
```
Comment (G6EJD) [inline examples/Waveshare_7_5/Waveshare_7_5.ino]: OK, but we could use just one char array. output[] but I want to retain capacity for 'Sat Apr-20-2019 12:31:45PM' which is what some US users prefer, so the minimum would be output[28], which is where the [3] came from as I added 2 for good measure in-case some adds for other formatting codes. My preference would be:
char   output[30];
And to change 'day_output' to 'output' to re-use the same array.

File: examples/Waveshare_7_5/Waveshare_7_5.ino [hunk 2]
```diff
@@ -498,58 +499,38 @@ boolean SetupTime() {
...
+        }
+    strftime(update_time, 9, "%H:%M:%S", &timeinfo);  // Creates: '@ 14:05:49'   and change from 30 to 8 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
+    sprintf(output, "%s %s", TXT_UPDATED, update_time);
   }
   else
   {
     strftime(day_output, 30, "%a %b-%d-%y", &timeinfo); // Creates  'Sat May-31-19'
-    strftime(output, 30, "( Updated: %r )", &timeinfo); // Creates: '@ 2:05:49pm'
-  }
+    strftime(output, 23, "Updated: %r", &timeinfo); // Creates: '@ 2:05:49pm'
```
Comment (G6EJD) [inline examples/Waveshare_7_5/Waveshare_7_5.ino]: good changes, but need a little more information. have you tested it between 'M' and 'I' modes?

File: examples/Waveshare_7_5/Waveshare_7_5.ino (hunk 1 above)
Comment (G6EJD) [reply examples/Waveshare_7_5/Waveshare_7_5.ino]: But, having thought about this some more, I need to be convinced that the tm structure is working across all Arduino IDE versions, for example it did not work (issues) with 1.0.0, seems OK with 1.0.2 and 1.0.3rc-1, but not everyone ahs the latest IDE. Needs testing to confirm.

**Comment by hoffmakl1961** [comment]
Only tested with "M".

File: examples/Waveshare_7_5/Waveshare_7_5.ino [hunk 3]
```diff
@@ -498,58 +499,38 @@ boolean SetupTime() {
   return TimeStatus;
 }
 //#########################################################################################
+// required an update of all lang_xx.h files
```
[the remark this answers is not in the document]
Comment (hoffmakl1961) [reply examples/Waveshare_7_5/Waveshare_7_5.ino]: the difference is that weekday and month will be defined as array. 
like in the lang_gr.h

//Wochentag
/*
```
const char* weekday_D[] = { "Sonntag",
                             "Montag",   
                             "Dienstag", 
                             "Mittwoch", 
                             "Donnerstag", 
                             "Freitag", 
                             "Samstag"};
*/
const char* weekday_D[] =  { "So.",
                             "Mo.",   
                             "Di.", 
                             "Mi.", 
                             "Do.", 
                             "Fr.", 
"Sa."};

//Monat
const char* month_M[] = { "Jan.",
                           "Feb.",
                           "März",
                           "Apr.",
                           "Mai",
                           "Juni",
                           "Juli",
                           "Aug.",
                           "Sep.",
                           "Okt.",
                           "Nov.",
                           "Dez."};
```

File: examples/Waveshare_7_5/Waveshare_7_5.ino [hunk 4]
```diff
@@ -67,7 +67,8 @@ boolean LargeIcon     = true, SmallIcon = false, RxWeather = false, RxForecast =
 #define Large  15           // For icon drawing, needs to be odd number for best effect
 #define Small  5            // For icon drawing, needs to be odd number for best effect
 String  time_str, date_str; // strings to hold time and received weather data;wi
-int     wifi_signal, CurrentHour = 0, CurrentMin = 0, CurrentSec = 0;
+int     wifi_signal, 
+uint8_t CurrentHour = 0, CurrentMin = 0, CurrentSec = 0;
```
[the remark this answers is not in the document]
Comment (hoffmakl1961) [reply examples/Waveshare_7_5/Waveshare_7_5.ino]: o.k.

File: examples/Waveshare_7_5/Waveshare_7_5.ino (hunk 1 above)
Comment (hoffmakl1961) [reply examples/Waveshare_7_5/Waveshare_7_5.ino]: Is it correct, you will use only one array? 
In the current code 
output will only be used für "updated: hh:mm:ss am" or pm ?

**Comment by G6EJD** [comment]
Show have you got a full change to propose? As I don't see a revised lang_gr.h and revised UpdateLocalTime version?

**Comment by hoffmakl1961** [comment]
I've moved the  UpdateLocalTime() by the following Code: 
```

//#########################################################################################
boolean UpdateLocalTime() {
  struct tm timeinfo;
  char   output[23], day_output[30];
  char   update_time[9];            // change from 30 to 9      
  while (!getLocalTime(&timeinfo, 5000)) {    // Wait for 5-sec for time to synchronise
    Serial.println("Failed to obtain time");
    return false;
  }
  CurrentHour = timeinfo.tm_hour;
  CurrentMin  = timeinfo.tm_min; 
  CurrentSec  = timeinfo.tm_sec;  
  
  //See http://www.cplusplus.com/reference/ctime/strftime/
  //Serial.println(&timeinfo, "%a %b %d %Y   %H:%M:%S");      // Displays: Saturday, June 24 2017 14:05:49
  Serial.println(&timeinfo, "%H:%M:%S");                      // Displays: 14:05:49
  if (Units == "M") {
    if (Language == "DE"){      
      sprintf(day_output,"%s, %02u. %s %04u",weekday_D[timeinfo.tm_wday],timeinfo.tm_mday,month_M[timeinfo.tm_mon],(timeinfo.tm_year)+1900); 
      // day_output >> So., 23. Juni 2019 <<   
      } 
      else 
      { sprintf(day_output,"%s %02u-%s-%04u",weekday_D[timeinfo.tm_wday],timeinfo.tm_mday,month_M[timeinfo.tm_mon],(timeinfo.tm_year)+1900);
        }
    strftime(update_time, 9, "%H:%M:%S", &timeinfo);  // Creates: '@ 14:05:49'   and change from 30 to 8 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
    sprintf(output, "%s %s", TXT_UPDATED, update_time);
  }
  else
  {
    strftime(day_output, 30, "%a %b-%d-%y", &timeinfo); // Creates  'Sat May-31-19'
    strftime(output, 23, "Updated: %r", &timeinfo); // Creates: '@ 2:05:49pm'
  }
  date_str = day_output;
  Serial.println(day_output);
  time_str = output;
  Serial.println(output);  
  return true;
}

```
The code  required a new lang_xx.h for weeksday and month.
as defined for DE.

```

//Temperatur - Luftfeuchtigkeit - Vorschau
const String TXT_FORECAST_VALUES  = "3 Tages Vorhersage";
const String TXT_CONDITIONS       = "Aktuelles Wetter";
const String TXT_DAYS             = "(Tage)";
const String TXT_TEMPERATURES     = "Temperatur";
const String TXT_TEMPERATURE_C    = "Temperatur [°C]";
const String TXT_TEMPERATURE_F    = "Temperatur [°F]";
const String TXT_HUMIDITY_PERCENT = "Luftfeuchte [%]";

//Luftdruck
const String TXT_PRESSURE         = "Luftdruck";
const String TXT_PRESSURE_HPA     = "Luftdruck [hPa]";
const String TXT_PRESSURE_IN      = "Luftdruck [in]";
const String TXT_PRESSURE_STEADY  = "beständig";
const String TXT_PRESSURE_RISING  = "steigend";
const String TXT_PRESSURE_FALLING = "fallend";

//Niederschlag / Schneefall
const String TXT_RAINFALL_MM = "Niederschlag [mm]";
const String TXT_RAINFALL_IN = "Niederschlag [in]";
const String TXT_SNOWFALL_MM = "Niederschlag [mm]";
const String TXT_SNOWFALL_IN = "Niederschlag [in]";
const String TXT_PRECIPITATION_SOON = "Regen/Schnee";
const String TXT_RAIN               = " Regen";
const String TXT_SNOW               = " Schnee";
const String TXT_EQUAL_RAIN         = " = Regen";
const String TXT_EQUAL_SNOW         = " = Schnee";

//Sonne
const String TXT_SUNRISE  = "Sonnenaufgang";
const String TXT_SUNSET   = "Sonnenuntergang";

//Mond
const String TXT_MOON_NEW             = "Neumond";          // Neumond
const String TXT_MOON_WAXING_CRESCENT = "zunehmende  Sichel";   // erstes Viertel bzw. zunehmende Sichel
const String TXT_MOON_FIRST_QUARTER   = "zunehmender Halbmond";   // zunehmender Halbmond 
const String TXT_MOON_WAXING_GIBBOUS  = "zweites Viertel";      // zweites Viertel
const String TXT_MOON_FULL            = "Vollmond";         // Vollmond
const String TXT_MOON_WANING_GIBBOUS  = "drittes Viertel";      // drittes Viertel 
const String TXT_MOON_THIRD_QUARTER   = "abnehmender Halbmond";   // abnehmender Halbmond
const String TXT_MOON_WANING_CRESCENT = "abnehmende Sichel";    // letztes Viertel bzw. abnehmende Sichel 

//Energie / WLAN
const String TXT_POWER  = "Akku";
const String TXT_WIFI   = "WLAN";
const char* TXT_UPDATED = "aktualisiert:";


//Wind
const String TXT_WIND_SPEED_DIRECTION = "Wind Geschw./Richtung";
const String TXT_N   = "N";
const String TXT_NNE = "NNO";
const String TXT_NE  = "NO";
const String TXT_ENE = "ONO";
const String TXT_E   = "O";
const String TXT_ESE = "OSO";
const String TXT_SE  = "SO";
const String TXT_SSE = "SSO";
const String TXT_S   = "S";
const String TXT_SSW = "SSW";
const String TXT_SW  = "SW";
const String TXT_WSW = "WSW";
const String TXT_W   = "W";
const String TXT_WNW = "WNW";
const String TXT_NW  = "NW";
const String TXT_NNW = "NNW";

//Wochentag
/*
const char* weekday_D[] = { "Sonntag",
                             "Montag",   
                             "Dienstag", 
                             "Mittwoch", 
                             "Donnerstag", 
                             "Freitag", 
                             "Samstag"};
*/
const char* weekday_D[] =  { "So.",
                             "Mo.",   
                             "Di.", 
                             "Mi.", 
                             "Do.", 
                             "Fr.", 
"Sa."};

//Monat
const char* month_M[] = { "Jan.",
                           "Feb.",
                           "März",
                           "Apr.",
                           "Mai",
                           "Juni",
                           "Juli",
                           "Aug.",
                           "Sep.",
                           "Okt.",
                           "Nov.",
                           "Dez."};
```


```

**Comment by G6EJD** [comment]
OK, I've made the changes now in a new version, just uploaded v16.2, lot's of changes plus yours. You can edit your language file locally to your taste/requirement.

**Pull request closed without merging**